### PR TITLE
misc: improve unit test coverage

### PR DESCRIPTION
**Pull request by rokusk**
## This PR adds tests to make sure all branches are covered.

For SafeMath library, there're a few lines that's not reachable, so need to use a `SafeMathTest` contract to call those line directly. 

Another interesting edge case is if a user initiates a small withdraw amount, and after the round settles, the value of the shares  = 0, it should fail instead of trying to transfer amount 0.

### CODE REVIEW

**Review by joshua0702k — CHANGES_REQUESTED** [review]

File: contracts/core/BaseVault.sol [hunk 1]
```diff
@@ -296,7 +296,7 @@ contract BaseVault is ReentrancyGuard, Ownable, ERC20, Initializable {
     // This checks if there is a withdrawal
```
[the remark this answers is not in the document]
Comment (rokusk) [reply contracts/core/BaseVault.sol:296]: The `_transfer` in `_redeem` is to transfer the shares the vault owes the user, so it's paying shares to user from vault; 

where as the `_transfer` in `initiateWithdraw` is to make sure users lock up their shares, so it's pulling shares from user to the vault. 

An easier way to think about it is that, they are 2 separated actions, calling `_redeem` as the first step in `initiateWithdraw` is not necessary, but it improve the UX: if user deposited in the middle of the round before, it can initiate a withdraw without needing to claim their shares first.

File: test/unit-tests/core/vault.ts [hunk 2]
```diff
@@ -133,11 +136,17 @@ describe('Unit test: Basic LyraVault flow', async () => {
     it('should revert when trying to set a performance fee that\'s too high', async() => {
```
[the remark this answers is not in the document]
Comment (rokusk) [reply test/unit-tests/core/vault.ts:136]: good catch. It requires a bit more thinking and potential refactoring, so I'll open another branch on top of this one to fix it.

[the remark this answers is not in the document]
Comment (rokusk) [reply test/unit-tests/core/vault.ts:136]: issues: #14

File: test/unit-tests/core/vault-shares.ts [hunk 3]
```diff
@@ -150,6 +150,13 @@ describe('Unit test: share calculating for pending deposit and withdraw', async
         const sharesToWithdraw = depositAmount
         await expect(vault.connect(depositor).redeem(sharesToWithdraw)).to.be.revertedWith('Exceeds available')
       })
+
+      it('should get 0 share if calling maxRedeem with no deposit Receipt', async() => {
+        const sharesBefore = await vault.balanceOf(anyone.address)
+        await vault.connect(anyone).maxRedeem()
+        const sharesAfter = await vault.balanceOf(anyone.address)
+        expect(sharesAfter).to.be.eq(sharesBefore)
+      })
```
[the remark this answers is not in the document]
Comment (rokusk) [reply test/unit-tests/core/vault-shares.ts:162]: Let me know if I'm following the same logic: 
```typescript
 171 | it('should return 0 for shares before round ends', async() => {
        const balances = await vault.shareBalances(depositor.address)
        expect(balances.heldByVault).to.be.eq(0)
        expect(balances.heldByAccount).to.be.eq(0)
  
        const shares = await vault.shares(depositor.address)
        expect(shares).to.be.eq(0)
      })
```
yes, pricePerShare is still 0 (we have no placeholder so it's 0), but L172 won't revert because In `getSharesFromReceipt`,  it only call `assetToShares` if the depositReceipt.round < currentRound, so in this test, it will just return `unredeemedShares` which is zero
```typescript
  if (depositReceipt.round > 0 && depositReceipt.round < currentRound) {
    uint sharesFromRound = assetToShares(depositReceipt.amount, assetPerShare, decimals);

    return uint(depositReceipt.unredeemedShares).add(sharesFromRound);
  }
 return depositReceipt.unredeemedShares;
```

[the remark this answers is not in the document]
Comment (rokusk) [reply test/unit-tests/core/vault-shares.ts:162]: I think they added this because people usually don't like view functions to revert, it could get quite annoying

File: test/unit-tests/core/vault-shares.ts [hunk 4]
```diff
@@ -326,7 +326,7 @@ describe('Unit test: share calculating for pending deposit and withdraw', async
         expect(vaultBalanceAfter.sub(vaultBalanceBefore)).to.be.eq(settlementPayout)
       })
       it('should revert if trying to completeWithdraw', async() => {
-        await expect(vault.connect(depositor).completeWithdraw()).to.be.revertedWith('Round not closed')
+        await expect(vault.connect(depositor).completeWithdraw()).to.be.revertedWith('Round in progress')
       })
       it('should rollover the vault to the next round', async() => {
         await vault.closeRound()
```
[the remark this answers is not in the document]
Comment (rokusk) [reply test/unit-tests/core/vault-shares.ts:332]: make sense, will add this one

File: contracts/core/BaseVault.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshua0702k) [reply contracts/core/BaseVault.sol:296]: Makes sense, thanks for explaining that.

File: test/unit-tests/core/vault-shares.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshua0702k) [reply test/unit-tests/core/vault-shares.ts:162]: Sorry for the confusion but I was using the wrong code line -> I'm referring to L180
`      it('should return constant for price per share before first share is minted', async() => {
        const oneShare = parseEther('1')
        const price = await vault.pricePerShare()
        expect(price).to.be.eq(oneShare)
      })`

[the remark this answers is not in the document]
Comment (rokusk) [reply test/unit-tests/core/vault-shares.ts:162]: it's because in `VaultLifecycle.rollover`
```typescript
50|  uint currentBalance = IERC20(asset).balanceOf(address(this));

     newPricePerShare = ShareMath.pricePerShare(currentShareSupply, currentBalance, pendingAmount, decimals);

     uint _mintShares = ShareMath.assetToShares(pendingAmount, newPricePerShare, decimals);
```
They're using the `currentShareSupploy` to calculate `newPricePerShare`, nad then using that to calculate how much shares (`_mintShares`) to mint.
 
if the current supply is 0, price should be default to 1, means you get exact shares based on your deposit. Otherwise _mintShares will be infinity (and causing the revert, cause price per share is 0).

That's why it should return 1 when there's no "shares" minted. not necessary when no assets has been deposited.

[the remark this answers is not in the document]
Comment (joshua0702k) [reply test/unit-tests/core/vault-shares.ts:162]: Makes sense - in short, the first depositor should get "1" as price per share since no fee was taken and no profit was made yet

**Review by joshua0702k — APPROVED** [review]
